### PR TITLE
KOGITO-2158: Review and fix kogito examples using persistence profile

### PR DESCRIPTION
**Pull request by Sgitario**
JIRA: https://issues.redhat.com/browse/KOGITO-2158
Objectives:
- Normalize all the examples that require Infinispan for the tests

### CODE REVIEW

File: common-example-utils/src/main/java/org/kie/kogito/testcontainers/ConditionalGenericContainer.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+import java.util.Optional;
+
+import org.apache.commons.lang3.StringUtils;
+import org.testcontainers.containers.GenericContainer;
+
+/**
+ * Container to be run if and only if it was enabled.
+ */
+public abstract class ConditionalGenericContainer<SELF extends GenericContainer<SELF>> extends GenericContainer<SELF> {
```
Comment (Sgitario) [inline common-example-utils/src/main/java/org/kie/kogito/testcontainers/ConditionalGenericContainer.java]: I created this abstraction because if the PR is approved, I'm planning to follow the same approach for Kafka. So, Kafka would reuse the same mechanism to enable/disable the container.

File: onboarding-example/onboarding-quarkus/pom.xml [hunk 2]
```diff
@@ -125,6 +130,20 @@
...
+            <artifactId>maven-surefire-plugin</artifactId>
+            <configuration>
+              <systemPropertyVariables>
+                <tests.category>persistence</tests.category>
+                <container.image.infinispan>${container.image.infinispan}</container.image.infinispan>
+              </systemPropertyVariables>
+            </configuration>
+          </plugin>
+        </plugins>
+      </build>
```
Comment (Sgitario) [inline onboarding-example/onboarding-quarkus/pom.xml:146]: Integration tests should use failsafe plugin instead of surefire (which should be used for unit tests only). However, I didn't want to do any unrelated changes as part of this PR.

File: process-infinispan-persistence-springboot/src/test/java/org/acme/deals/DealsRestIT.java [hunk 3]
```diff
@@ -15,46 +15,32 @@
...
-import static org.junit.jupiter.api.Assertions.assertNotNull;
+import io.restassured.RestAssured;
+import io.restassured.http.ContentType;
 
 @Testcontainers
-@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT, properties = {
-        "infinispan.remote.use-auth=true",
-        "infinispan.remote.auth-username=admin",
-        "infinispan.remote.auth-password=admin",
-        "infinispan.remote.sasl-mechanism=DIGEST-MD5"
```
Comment (Sgitario) [inline process-infinispan-persistence-springboot/src/test/java/org/acme/deals/DealsRestIT.java:46]: I changed the sasl-mechanism to "PLAIN" as some tests were failing using DIGEST (a missing property probably). I guess using "PLAIN" should be fine.

File: common-example-utils/README.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+```
+# Infinispan
+infinispan.remote.server-list=localhost:11222
+infinispan.remote.sasl-mechanism=PLAIN
+infinispan.remote.auth-server-name=infinispan
+infinispan.remote.use-auth=true
+infinispan.remote.auth-realm=default
+infinispan.remote.auth-username=admin
+infinispan.remote.auth-password=admin
```
Comment (Sgitario) [inline common-example-utils/README.md]: I set the SASL mechanism to "PLAIN" and added the "AUTH-REALM" to "default" as some tests did need it to work.

File: process-springboot-example/src/test/java/org/kie/kogito/examples/demo/PersonsRestTest.java [hunk 5]
```diff
@@ -32,22 +32,30 @@
...
+@Testcontainers
 @SuppressWarnings("rawtypes")
 @ExtendWith(SpringExtension.class)
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT, classes = DemoApplication.class)
 @DirtiesContext(classMode = ClassMode.AFTER_EACH_TEST_METHOD) // reset spring context after each test method
 public class PersonsRestTest {
 	
+    @Container
+    public static GenericContainer<?> INFINISPAN = new InfinispanContainer().enableIfTestCategoryIs("persistence");
+
```
Comment (Sgitario) [inline process-springboot-example/src/test/java/org/kie/kogito/examples/demo/PersonsRestTest.java:58]: I see that this is using TABs, that's why this format issue. Should I reformat the whole file? (I will result in much more changes...)

File: process-springboot-example/src/test/java/org/kie/kogito/examples/demo/OrdersProcessTest.java [hunk 6]
```diff
@@ -42,12 +43,19 @@
...
 
+@Testcontainers
 @ExtendWith(SpringExtension.class)
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT, classes = DemoApplication.class)
 @DirtiesContext(classMode = ClassMode.AFTER_EACH_TEST_METHOD) // reset spring context after each test method
 public class OrdersProcessTest {
 
+    @Container
+    public static GenericContainer<?> INFINISPAN = new InfinispanContainer().enableIfTestCategoryIs("persistence");
+
```
Comment (Sgitario) [inline process-springboot-example/src/test/java/org/kie/kogito/examples/demo/OrdersProcessTest.java:58]: I see that this is using TABs, that's why this format issue. Should I reformat the whole file? (I will result in much more changes...)

**Comment by Sgitario** [comment]
I ran successfully the whole examples without an external Infinispan instance and all the tests passed.

File: common-example-utils/pom.xml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    </dependency>
+    
+    <dependency>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-junit5</artifactId>
+      <scope>provided</scope>
+    </dependency>
+  </dependencies>
+  
+  <dependencyManagement>
```
Comment (cristianonicolai) [inline common-example-utils/pom.xml]: please move _dependencyManagement_ to before _dependencies_ for better readbility

File: onboarding-example/onboarding-quarkus/pom.xml (hunk 2 above)
Comment (cristianonicolai) [reply onboarding-example/onboarding-quarkus/pom.xml:146]: @jstastny-cz should this be a follow-up PR or you happy to be fixed in here.
either way is fine for me.

File: process-springboot-example/src/test/java/org/kie/kogito/examples/demo/OrdersProcessTest.java (hunk 6 above)
Comment (cristianonicolai) [reply process-springboot-example/src/test/java/org/kie/kogito/examples/demo/OrdersProcessTest.java:58]: fine as is, probably is best to do a new PR with just the reformat after merging this.

File: common-example-utils/README.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
+# Common Utils For Kogito Examples
+
+## Infinispan Test Containers Support
+
+Add the common utils dependency in the _pom.xml_ file:
+
+```xml
+<dependency>
+  <groupId>org.kie.kogito</groupId>
+  <artifactId>common-example-utils</artifactId>
```
Comment (cristianonicolai) [inline common-example-utils/README.md]: perhaps we should be more clear that this is for test only? Perhaps name it _examples-test-common_ ?

File: common-example-utils/pom.xml (hunk 7 above)
Comment (Sgitario) [reply common-example-utils/pom.xml]: done

File: common-example-utils/README.md (hunk 8 above)
Comment (Sgitario) [reply common-example-utils/README.md]: Renamed to "examples-test-utils"

File: onboarding-example/onboarding-quarkus/pom.xml (hunk 2 above)
Comment (jstastny-cz) [reply onboarding-example/onboarding-quarkus/pom.xml:146]: I don't mind either. I'd just create Jira for this and wrap this up quickly. @Sgitario do you have one? If not, please create and link the issue in this comment thread for future reference.

**Review by jstastny-cz — APPROVED** [review]

**Review by sutaakar — APPROVED** [review]
Looks good to me.

File: onboarding-example/onboarding-quarkus/pom.xml (hunk 2 above)
Comment (Sgitario) [reply onboarding-example/onboarding-quarkus/pom.xml:146]: Raised https://issues.redhat.com/browse/KOGITO-2467

**Comment by Sgitario** [comment]
jenkins please retest this

**Review by cristianonicolai — APPROVED** [review]
